Initializing the business logic layer

IThror10 opened the pull request.

IThror10 committed:
[FR #29] Create DTO Classes && Base Exception class

In order to transfer data to and from DB and WEB layers,
the necessary DTO classes and base exception class
were added to the BL layer.

Further custom exception classes will inherit
from the created `ProcessException` class.

To make some DTO classes more comfortable to use,
auxiliary constructors have been added.

IThror10 committed:
[FR #13] Create Command Classes && Command Visitor

Each highlighted user action corresponds to a certain `*Command` class.
This class implements the Visitor pattern. Further logic of
interaction with objects of `*Command` classes will be implemented
exactly in the Visitor class.

Further adding commands requires creating a new class that is
a descendant of `ACommand` and adding the appropriate method
to the CommandVisitor

IThror10 committed:
[FR #12] Create Query Classes && Query Visitor

Each highlighted user query corresponds to a certain `*Query` class.
This class implements the Visitor pattern. Further logic of
interaction with objects of `*Query` classes will be implemented
exactly in the Visitor class.

Further adding queries requires creating a new class that is
a descendant of `AQuery` and adding the appropriate method
to the QueryVisitor.

aavortep commented on portfolio/src/main/scala/ru/bmstu/portfolio/dto/User.scala:
```diff
@@ -0,0 +1,27 @@
+)
+
+sealed case class UserPublicData
+(
```
Единственный вопрос. А где пароль будет храниться? Или это у тебя как раз токен в классе выше?

aavortep replied:
@IThror10

IThror10 replied:
Это DTO в BL. То что будет возвращаться по запросу, когда пользователи будут искать информацию о других пользователях: `GET /users/{user_id}`

Информация хранится в БД. Этих табличек пока нет.

IThror10 committed:
[FR #14] Create abstract factory && TableGateways

As a result of the design, it was decided to interact with the DB
layer by using the TableGateway pattern. To control the belonging
of all created classes to a single family, the Abstract Factory
pattern was introduced.

TableGatewayFactory is an implementation of this pattern,
and is a Singleton object. Registration of factories is done
by calling its method `registerFactory`. A custom exception
class has been added for empty option errors.

To test Singleton's correctness, it was covered by unit tests.
To do this, the necessary dependencies were added to the build.sbt file.

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/command/ACommand.scala:
```diff
@@ -0,0 +1,10 @@
+import ru.bmstu.portfolio.dto.ProcessException
+
+abstract class ACommand[Result] {
+  val result: Option[Result] = None
```
Must be var value

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/command/album/AddCoauthorCommand.scala:
```diff
@@ -0,0 +1,9 @@
+import ru.bmstu.portfolio.dto.Album
+
+sealed case class AddCoauthorCommand(userId: Int, newUserId: Int, albumId: Int) extends ACommand[Album] {
+  override def acceptVisitor(visitor: TCommandVisitor): Unit =
```
throws Exception annotation

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/command/album/CreateAlbumCommand.scala:
```diff
@@ -0,0 +1,9 @@
+import ru.bmstu.portfolio.dto.Album
+
+sealed case class CreateAlbumCommand(userId: Int, addUserId: Int, albumId: Int) extends ACommand[Album] {
+  override def acceptVisitor(visitor: TCommandVisitor): Unit =
```
throws Exception annotation

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/command/album/DeleteAlbumCommand.scala:
```diff
@@ -0,0 +1,9 @@
+import ru.bmstu.portfolio.dto.CommandResult
+
+sealed case class DeleteAlbumCommand(albumId: Int, userId: Int) extends ACommand[CommandResult] {
+  override def acceptVisitor(visitor: TCommandVisitor): Unit =
```
Throws Exception there and in other queries and commands

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/factory/TableGatewayFactory.scala:
```diff
@@ -0,0 +1,36 @@
+import scala.collection.mutable
+
+trait TableGatewayFactory {
+  @throws(classOf[ProcessException])
```
wrong style annotation

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/query/AQuery.scala:
```diff
@@ -0,0 +1,10 @@
+abstract class AQuery[Result] {
+  val result: Option[Result] = None
+
+  @throws(classOf[ProcessException])
```
Wrong style annotation

IThror10 commented on portfolio/src/main/scala/ru/bmstu/portfolio/query/TQueryVisitor.scala:
```diff
@@ -0,0 +1,31 @@
+
+
+  @throws[ProcessException]
+  def visitGetWorksBuAlbumIdQuery(query: GetWorksByAlbumIdQuery): Unit
```
Change Bu to By

IThror10 reviewed (commented):
Исправить замечания

IThror10 committed:
[BF #31] Add @throws; Change val -> var in ACommand, AQuery

Added throws annotations for the acceptVisitor method of ACommand and AQuerydescendant
classes. Changed naming of @throws(classOf[Exception]) to @throws[Exception]

Fixed syntax errors in naming one of the QueryVisitor functions

In abstract classes ACommand and AQuery, variable type was changed from val to var

IThror10 committed:
[FR #16] Create Common Handler && Handler Visitors

A CommonHandler class was created to handle user queries and commands.
CommandHandlerVisitor and QueryHandlerVisitor are used to process
user commands and queries.

These classes are implementing the logic of interaction with
the database and saving the result of execution of command/query
in the appropriate command/query result field
